src/ImageSharp/Common/Helpers/SimdUtils.HwIntrinsics.cs

```diff
@@ -246,6 +312,75 @@ public static class HwIntrinsics
                 }
             }
 
+            [MethodImpl(InliningOptions.ShortMethod)]
+            private static unsafe void Pad3Shuffle4(
+                ReadOnlySpan<byte> source,
+                Span<byte> dest,
+                byte control)
+            {
+                if (Ssse3.IsSupported)
+                {
+                    Vector128<byte> wMask = Vector128.Create(0xff000000u).AsByte();
+                    Vector128<byte> padMask = Vector128.Create(0, 1, 2, -1, 3, 4, 5, -1, 6, 7, 8, -1, 9, 10, 11, -1).AsByte();
+
+                    Span<byte> bytes = stackalloc byte[Vector128<byte>.Count];
+                    Shuffle.MmShuffleSpan(ref bytes, control);
+                    Vector128<byte> vcm = Unsafe.As<byte, Vector128<byte>>(ref MemoryMarshal.GetReference(bytes));
+
+                    fixed (byte* sBase = source)
+                    fixed (byte* dBase = dest)
+                    {
+                        byte* s = sBase;
+                        byte* d = dBase;
+
+                        // TODO: Consider unrolling and shuffling 4 at a time using Ssse3.AlignRight
+                        // See https://stackoverflow.com/questions/2973708/fast-24-bit-array-32-bit-array-conversion
+                        for (int i = 0; i < source.Length; i += 16)
+                        {
+                            Vector128<byte> vs0 = Sse2.LoadVector128(s);
+                            Vector128<byte> val = Sse2.Or(wMask, Ssse3.Shuffle(vs0, padMask));
+                            val = Ssse3.Shuffle(val, vcm);
+                            Sse2.Store(d, val);
+
+                            s += 12;
+                            d += 16;
+                        }
+                    }
+                }
+            }
+
+            [MethodImpl(InliningOptions.ShortMethod)]
+            private static unsafe void Shuffle4Slice3(
+                ReadOnlySpan<byte> source,
+                Span<byte> dest,
+                byte control)
+            {
+                if (Ssse3.IsSupported)
+                {
+                    Vector128<byte> sliceMask = Vector128.Create(0, 1, 2, 4, 5, 6, 8, 9, 10, 12, 13, 14, -1, -1, -1, -1).AsByte();
+
+                    Span<byte> bytes = stackalloc byte[Vector128<byte>.Count];
+                    Shuffle.MmShuffleSpan(ref bytes, control);
+                    Vector128<byte> vcm = Unsafe.As<byte, Vector128<byte>>(ref MemoryMarshal.GetReference(bytes));
+
+                    fixed (byte* sBase = source)
+                    fixed (byte* dBase = dest)
+                    {
+                        byte* s = sBase;
+                        byte* d = dBase;
+
+                        for (int i = 0; i < source.Length; i += 16)
+                        {
+                            Vector128<byte> vs0 = Ssse3.Shuffle(Sse2.LoadVector128(s), vcm);
+                            Sse2.Store(d, Ssse3.Shuffle(vs0, sliceMask));
```

Comment:
This will overwrite the end of your output buffer by 4 bytes on the last store, and I don't see that you're stopping short on the input slice to account for that.  The version in the SO answer you linked  reads 4 full vectors and writes 3, so it'll be both safer and faster.  Where did you get stuck on that?

Reply:
Just had another look at that SO question... the assembly answer covers both directions, but it goes about the 32->24 shuffle a bit strangely.   I worked up a quick C# sample to show a simpler way of doing it:

```C#
unsafe static void narrow32to24()
{
    var arr = new byte[64];
    for (int i = 0; i < arr.Length; i += 4)
    {
        int j = i * 3 / 4;
        arr[i] = (byte)j;
        arr[i+1] = (byte)(j + 1);
        arr[i+2] = (byte)(j + 2);
        arr[i+3] = byte.MaxValue;
    }

    var vmasko = Vector128.Create(0, 1, 2, 4, 5, 6, 8, 9, 10, 12, 13, 14, 3, 7, 11, 15).AsByte();
    var vmaske = Ssse3.AlignRight(vmasko, vmasko, 12).AsByte();

    fixed (byte* bp = &arr[0])
    {
        var v0 = Sse2.LoadVector128(bp);
        var v1 = Sse2.LoadVector128(bp + Vector128<byte>.Count);
        var v2 = Sse2.LoadVector128(bp + Vector128<byte>.Count * 2);
        var v3 = Sse2.LoadVector128(bp + Vector128<byte>.Count * 3);

        v0 = Ssse3.Shuffle(v0, vmaske);
        v1 = Ssse3.Shuffle(v1, vmasko);
        v2 = Ssse3.Shuffle(v2, vmaske);
        v3 = Ssse3.Shuffle(v3, vmasko);

        v0 = Ssse3.AlignRight(v1, v0, 4);
        v3 = Ssse3.AlignRight(v3, v2, 12);

        v1 = Sse2.ShiftLeftLogical128BitLane(v1, 4);
        v2 = Sse2.ShiftRightLogical128BitLane(v2, 4);

        v1 = Ssse3.AlignRight(v2, v1, 8);

        Console.WriteLine(v0);
        Console.WriteLine(v1);
        Console.WriteLine(v3);
    }
}
```

And a crappy ascii diagram to show how it works.  The only tricky bit here is that I drew the diagram in memory order, but since Intel is little-endian, right and left mean the opposite on their shifts:

```
       v0               v1               v2               v3
BGRxBGRxBGRxBGRx BGRxBGRxBGRxBGRx BGRxBGRxBGRxBGRx BGRxBGRxBGRxBGRx
pshufb(maskeven) pshufb(maskodd)  pshufb(maskeven) pshufb(maskodd)
       ||               ||               ||               ||
       \/               \/               \/               \/
xxxxBGRBGRBGRBGR BGRBGRBGRBGRxxxx xxxxBGRBGRBGRBGR BGRBGRBGRBGRxxxx
palignr(v1, v0, 4)                                palignr(v3, v2, 12)
       ||                                                 ||
       \/                                                 \/
BGRBGRBGRBGRBGRB                                   RBGRBGRBGRBGRBGR
                 pslldq(v1, 4)    psrldq(v2, 4)
		        ||               ||
                        \/               \/
                 xxxxBGRBGRBGRBGR BGRBGRBGRBGRxxxx
                 palignr(v2, v1, 8)
                        ||
                        \/
                 GRBGRBGRBGRBGRBG

```

Reply:
I got stuck because it looks like they're actually working from `uint` for both source and output. I'll give your approach a go though thanks!

Reply:
Gonna have to see this in a loop. I'm doing something stupid.